src/array/ops/cast.rs

```diff
@@ -701,12 +705,120 @@ impl EmbeddingArray {
+                Python::with_gil(|py| {
+                    let mut ndarrays = Vec::with_capacity(self.len());
+                    let mut bitmap = arrow2::bitmap::MutableBitmap::with_capacity(self.len());
+                    for i in 0..self.len() {
+                        let img_buf = self.as_image_obj(i);
+                        let ndarray = img_buf
+                            .map(|buf| {
+                                let slice = buf.as_u8_slice();
+                                let shape = (
+                                    buf.height() as usize,
+                                    buf.width() as usize,
+                                    buf.mode().num_channels() as usize,
+                                );
+                                // NOTE(Clark): This will create a copy in the Python heap, using NumPy for allocation.
+                                PyArray::from_slice(py, slice).reshape(shape.into_dimension())
+                            })
+                            .transpose()
+                            .map_err(|e| {
+                                DaftError::ValueError(format!(
+                                    "Converting image to ndarray failed: {}",
+                                    e
+                                ))
+                            })?;
+                        ndarrays.push(
+                            ndarray
+                                .unwrap_or(PyArray::zeros(
+                                    py,
+                                    (1_usize, 1_usize, 1_usize).into_dimension(),
+                                    false,
+                                ))
+                                .deref()
+                                .to_object(py),
+                        );
+                        bitmap.push(ndarray.is_some());
+                    }
+                    let values_array = PseudoArrowArray::new(ndarrays.into(), bitmap.into());
+                    Ok(PythonArray::new(
+                        Field::new(self.name(), dtype.clone()).into(),
+                        values_array.to_boxed(),
+                    )?
+                    .into_series())
+                })
+            }
+            _ => self.physical.cast(dtype),
+        }
     }
 }
 
+#[cfg(feature = "python")]
+fn fixed_shape_image_array_to_python_array<
+    Tgt: numpy::Element + NumCast + ToPrimitive + arrow2::types::NativeType,
+>(
+    py: Python<'_>,
+    image_array: &FixedShapeImageArray,
+    dtype: &DataType,
+    height: &u32,
+    width: &u32,
+    mode: ImageMode,
+) -> DaftResult<PythonArray> {
+    let shape = (
+        *height as usize,
+        *width as usize,
+        mode.num_channels() as usize,
+    )
+        .into_dimension();
+    let mut ndarrays = Vec::with_capacity(image_array.len());
+    let mut bitmap = arrow2::bitmap::MutableBitmap::with_capacity(image_array.len());
+    for arr in image_array.as_arrow().iter() {
```

Comment:
The alternative here is that we convert the entire data array into numpy view (via the pyarrow route) and then slice it up for each row

Reply:
Should note that our FFI layer currently makes a full copy of fixed-size list arrays if they are slice views, to work around a pyarrow bug. https://github.com/Eventual-Inc/Daft/blob/6cdc4b5fa1920a229e94ee146137871d8f0ebcff/src/ffi.rs#L129

Shouldn't affect us in the common case (fixed-size list array isn't a slice view) though.

Reply:
To avoid a few of the inefficiencies (extra copies) mentioned, I pushed up a commit delegating to pyarrow's `.to_numpy()` conversion for each image, rather than converting the entire data array to a NumPy view and then slicing. What do you think about revisiting the creation of a top-level ndarray view for the full data array once we are exposing a top-level ndarray view in the egress APIs (e.g. the `Series.to_numpy()` API)?

Reply:
Here's what the top-level ndarray view + per-image slicing would look like, with the above caveats around the potential for extra copies:
```rust
impl FixedShapeImageArray {
    pub fn cast(&self, dtype: &DataType) -> DaftResult<Series> {
        match (dtype, self.logical_type()) {
            #[cfg(feature = "python")]
            (DataType::Python, DataType::FixedShapeImage(_, mode, height, width)) => {
                pyo3::Python::with_gil(|py| {
                    let shape = (
                        self.len(),
                        *height as usize,
                        *width as usize,
                        mode.num_channels() as usize,
                    );
                    let pyarrow = py.import("pyarrow")?;
                    let py_array = ffi::to_py_array(self.as_arrow().values().clone(), py, pyarrow)?
                        .call_method1(py, pyo3::intern!(py, "to_numpy"), (false,))?
                        .call_method1(py, pyo3::intern!(py, "reshape"), (shape,))?;
                    let mut ndarrays = Vec::with_capacity(self.len());
                    for i in 0..self.len() {
                        let image_array = if self.physical.is_valid(i) {
                            py_array.as_ref(py).get_item(i)?.to_object(py)
                        } else {
                            PyArray3::<u8>::zeros(
                                py,
                                (shape.1, shape.2, shape.3).into_dimension(),
                                false,
                            )
                            .deref()
                            .to_object(py)
                        };
                        ndarrays.push(image_array);
                    }
                    let values_array =
                        PseudoArrowArray::new(ndarrays.into(), self.as_arrow().validity().cloned());
                    Ok(PythonArray::new(
                        Field::new(self.name(), dtype.clone()).into(),
                        values_array.to_boxed(),
                    )?
                    .into_series())
                })
            }
            (_, _) => self.physical.cast(dtype),
        }
    }
}
```